### ISSUE TITLE
Improving nnet generalization to feature offsets

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#977]
I have thought of an efficient way to enhance the generalization of our models to additive feature offsets by adding randomly chosen shifts to the training features.  @vijayaditya, perhaps you can work with someone on this.  Or @pegahgh, if you want to do this and can do it soon, you can do it, as I know you have been interested in similar things in the past.

This is something that we could do when dumping egs for neural net training, including chain training, where in effect we dump egs for [say] 5 different versions of each utterance, differing by random feature shifts... but we won't physically dump all of those egs; instead we store special data with the egs that allows the differently shifted versions to be generated on the fly during training.

I'll try to make it concrete below.

Firstly, we work out a covariance for the feature-space CMN offsets, by taking the centered covariance of the speaker or pseudo-speaker means, and use this times some constant (probably a bit less than one, like 0.5, or maybe 1] to generate random CMN offsets to be applied to the features.  [Or possibly the offsets could have non-random means, to subtract a little of the speaker's existing offset from the global mean, thus keeping the same speaker-level covariance as before the perturbation... we can experiment with this.]

Each speaker we will have a certain number of 'versions' of that speaker, differing in random CMN offsets [e.g. num_cmn_offsets=5].  [Note: this will probably be done after we've done the 'max2' stuff so they won't be 'real' speakers in general, but parts of speakers.]   Note: there is no point in making num_cmn_offsets larger than the number of epochs that we use in neural net training, as we wouldn't have time to cycle through all the offsets otherwise.  The cmn_offsets will be randomly generated at some point and put in an archive indexed by speaker or (duplicated) per utterance, in a matrix with num-rows == num_cmn_offsets and num-cols == feature-dim, e.g. 40.  I have more details below on how we can store this info.

When we dump the egs we will only physically dump one version of the eg.  However, we will abuse the egs format slightly by adding an extra NnetIo member called 'offsets', which contains the cmn_offsets for that utterance.  It will contain the above-mentioned matrix of offsets [dump it with compression]; and for the indexes, I suggest to have n=0, t=0, and x corresponding to each row of the matrix, so x=0, x=1, .. x=4 in the normal case... this is not critical at all, it just means that the Check() function won't complain.

If cmn_offsets are used, then the first thing that will be done to the egs while training [before nnet3-merge-egs] is to select one of the offsets and add it to the features.  This could be done by an extra option to nnet3-copy-egs, e.g. a --select-feature-offset=N option [this value N would be increased on each epoch, probably, modulo num_cmn_offsets].  The --select-feature-offset would add the chosen offset to the features, and would also select the iVectors that were generated for that the features offset by that value.  What I am suggesting below is to dump the iVectors all appended together, with a dimension equal to the iVector dim times num_cmn_offsets.

I now explain some details of how this interacts with the iVector extraction.
The feature offset does not affect the iVector simply via a shift, because of interaction with the prior term and [if we are near utterance begin], small effects due to non-invariance of the online CMN process to feature shift (c.f. the --global-cmvn-stats option).

I think the best way to proceed for now to actually extract different versions of the iVectors for each utterance, one for each cmn_offset.  In order to reduce excessive I/O or computation while extracting iVectors, it could be done as follows:

ivector-extract-online2.cc would be modified to accept a table of speaker-specific feature offsets, via a --spk2cmn-offsets option, reading a matrix of offsets per speaker.  When this code path is activated, each offset version of the utterance would be processed sequentially, cycling through the rows of the offset matrix, and the iVectors extracted for each offset version of the feature would be appended together before being output.  One code change required is that there would need to be a vector of  OnlineIvectorExtractorAdaptationState with dimension equal to the num_cmn_offsets; this can be obtained from the num-rows of the matrix read from spk2cmn_offsets_reader.  [this is so that we remember the state from previous utterances of the _same offset_ of the speaker].
This will obviously slow down the training-set iVector extraction by a factor of 5, but we can compensate to some extent by using e.g. twice the number of jobs. 

nnet3-get-egs would need to be extended by adding an --utt2cmn-offsets option (it would map to the matrix of offsets, one offset per row).  As mentioned above, nnet3-get-egs would create a new NnetIo object in the eg, called 'offsets'; and the iVectors it dumps will just have a dimension that is larger by a factor of 'num_cmn_offsets'.  [All nnet3-get-egs would have to do here is to check that the iVector dimension is a multiple of num_cmn_offsets, as the iVectors it is reading in will already have the larger dimension.]

Probably in nnet-utils.h you will declare a function like:

SelectFeatureOffset(int32 feature_offset,  NnetExample *eg);

[obviously with suitable documentation about what it does].
This function is expected to operate on single egs (egs with a single 'n' index), not on 'merged' egs that come from nnet3-merge-egs (for merged egs, the implementation would be more complex)... you can just assert that all the 'n' values equal zero.  Because this will be done as part of a pipeline and the results not written to disk, it's safe to uncompress any features that you have to modify, and not re-compress them.  You'll want to remove the 'offsets' element from the NnetIo object after processing the features in this way.

It will likely be the most convenient, when you modify nnet3-copy-egs.cc, to add an extra argument feature_offset to SelectFromExample() that is responsible for doing this.  If -1, this would mean there is no feature selection being done.

I suggest that you dump the validation and train-subset egs in the same way as the training-set egs (i.e. with the feature offsets), but, when you use them, if you are using features offsets, always use --select-feature-offset=0 for consistency, so that the diagnostics don't jump around from epoch to epoch.

Note: in order to avoid adding too many binaries, you could write a single program that computes the variance of the CMNs over the training speakers and then writes out the table of cmn_offsets per speaker, e.g.

```
    generate-random-cmn-offsets  <cmn-archive-in> <cmn-offset-archive-out>
 e.g.:
    generate-random-cmn-offsets --srand=0 --num-cmn-offsets=5 --cmn-offset-scale=0.5 data/train/feats.scp ark:cmn_offsets_per_spk.ark,cmn_offsets_per_spk.scp 
```

... this could work out the covariance of offsets from the training speakers' stats and use the same speaker list to generate the output.  The offset-scale would be applied to the generated offsets (so it's a scale on the stddev, not on the covariance).
  [Note: you could add a boolean option --preserve-total-covariance, which, if true, would, modify the CMN offsets to subtract some of the existing offset of the speaker in order to preserve the total covariance of speakers.  Suppose we've generated the random offset 'o' according to the centered covariance of the speaker means; and suppose that this speaker's mean is 'm', and the global mean is 'g'.  Then the code might look something like:

```
     if (!preserve_total_covariance) {
        offset = o * cmn_offset_scale;
    } else {
        // The factor 1.0 /  sqrt(1.0 + cmn_offset_scale * cmn_offset_scale)
        // ensures that by adding these offsets, we don't change the total covariance of
        // the speaker means after adding the offsets.  Note that if we replace
        // the sqrt factor with 1.0 below, it reduces to the case where preserve_total_covariance
        // is false.
        offset_from_global_mean = (o * cmn_offset_scale + (m - g)) / sqrt(1.0 + cmn_offset_scale * cmn_offset_scale);
        offset = offset_from_global_mean - (m - g);
    }
```

It may be most convenient to make cmn_offsets.scp part of the feature directory e.g. data/train_sp/cmn_offsets.scp [indexed by utterance.. this makes it easier to copy across different data-dirs that have different speaker-info, which we use in the standard scripts], and to have the scripts that deal with these offsets (e.g. steps/nnet3/get_egs.sh and steps/online/nnet2/extract_ivectors_online.sh) detect the presence of cmvn_offsets.scp and automatically do the right thing.  Obviously copy_data_dir.sh, fix_data_dir.sh, validate_data_dir.sh and split_data_dir.sh would have to know about it, but this is just question of adding it to lists of other utterance-indexed files such as utt2lang.  When it's needed per speaker, as by ivector-extract-online2, an inline command like awk '{print $1,$2}' <$split_feats/spk2utt | utils/apply_map.pl -f 2 $split_feats/cmn_offsets.scp can be used to obtain a speaker-indexed version of the file. 

You will need to be a bit careful with scripts that work out the iVector dim from the archive of iVectors, and with iVector dims in 'info' directories.  I think that generally speaking we should aim to have 'ivector_dim' variables in scripts and in info/ directories correspond to the "real" iVector dimension, which may not equal the dimensions of iVectors stored on disk; and where necessary add a 'num_cmn_offsets' variable or file.
